### ISSUE TITLE
WikiData Not Returning CSV

### ISSUE THREAD

**Issue by hendursaga** [RDFLib/sparqlwrapper#226]
While I can request WikiData to return JSON easily on SELECT queries, I cannot do the same for CSV/TSV, it just returns XML and then sparqlwrapper complains it's not in a format it's expecting.

I think the problem has to do with setting URL parameters like `format` and maybe `results` or `output`. See [this article](https://dirk.net/2021/03/22/returning-csv-from-wikidata-query-api/) and its [referenced link](https://www.mediawiki.org/wiki/Wikidata_Query_Service/User_Manual#Supported_formats).

I'm not sure if that's correct behavior or not on WikiData's side, but it would be nice to have at least a workaround.

**Comment by inormann** [comment]
Indeed, the problem seems to be that `format` parameter added by SPARQLWrapper. Fortunately the class comes with a method [`setOnlyConneg`](https://sparqlwrapper.readthedocs.io/en/latest/SPARQLWrapper.Wrapper.html#SPARQLWrapper.Wrapper.SPARQLWrapper.setOnlyConneg) to suppress these parameters and to rely on content negotiation only.

This seems to be the solution as this code snippet may demonstrate:

```
from SPARQLWrapper import SPARQLWrapper, CSV

sparql = SPARQLWrapper("https://query.wikidata.org/sparql", returnFormat=CSV)
sparql.setOnlyConneg(True) # enforce content negotiation only!
sparql.setQuery('SELECT * WHERE {?s ?p ?o} LIMIT 3')
ret = sparql.query()
ret._get_responseFormat() # -> 'csv'

```

**Comment by dayures** [comment]
Thanks for reporting this. I can reproduce the behavior, but this is endpoint negotiation behavior rather than a SPARQLWrapper parsing bug.

For `SELECT` with CSV/TSV, WikiData may ignore/override the `format`/`output`/`results` URL params and return XML unless content negotiation is used strictly. SPARQLWrapper supports a workaround as @inormann mentioned.:

```python
from SPARQLWrapper import SPARQLWrapper, CSV, TSV

sparql = SPARQLWrapper("https://query.wikidata.org/sparql")
sparql.setQuery("SELECT * WHERE { ?s ?p ?o } LIMIT 10")
sparql.setReturnFormat(CSV)          # or TSV
sparql.setOnlyConneg(True)           # important: rely only on Accept header
# optional hard override if needed:
# sparql.addCustomHttpHeader("Accept", "text/csv")

data = sparql.query().convert()      # bytes
print(data.decode("utf-8"))
```

Because this has a working client-side workaround and appears to be service-specific content negotiation behavior, I’m closing this issue. If someone can still reproduce with current SPARQLWrapper and `setOnlyConneg(True)`, please open a new issue with a minimal script + full response headers.
